apps/omg_watcher_info/lib/omg_watcher_info/db/txoutput.ex

```diff
@@ -79,23 +82,38 @@ defmodule OMG.WatcherInfo.DB.TxOutput do
     )
   end
 
-  def get_utxos(owner) do
-    query =
-      from(
-        txoutput in __MODULE__,
-        preload: [:ethevents],
-        left_join: ethevent in assoc(txoutput, :ethevents),
-        # select txoutputs by owner that have neither been spent nor have a corresponding ethevents exit events
-        where: txoutput.owner == ^owner and is_nil(txoutput.spending_txhash) and (is_nil(ethevent) or fragment("
- NOT EXISTS (SELECT 1
-             FROM ethevents_txoutputs AS etfrag
-             JOIN ethevents AS efrag ON
-                      etfrag.root_chain_txhash_event=efrag.root_chain_txhash_event
-                      AND efrag.event_type IN (?)
-                      AND etfrag.child_chain_utxohash = ?)", "standard_exit", txoutput.child_chain_utxohash)),
-        order_by: [asc: :blknum, asc: :txindex, asc: :oindex]
-      )
+  defp query_get_utxos(address) do
+    from(
+      txoutput in __MODULE__,
+      preload: [:ethevents],
+      left_join: ethevent in assoc(txoutput, :ethevents),
+      # select txoutputs by owner that have neither been spent nor have a corresponding ethevents exit events
+      where: txoutput.owner == ^address and is_nil(txoutput.spending_txhash) and (is_nil(ethevent) or fragment("
+NOT EXISTS (SELECT 1
+           FROM ethevents_txoutputs AS etfrag
+           JOIN ethevents AS efrag ON
+                    etfrag.root_chain_txhash_event=efrag.root_chain_txhash_event
+                    AND efrag.event_type IN (?)
+                    AND etfrag.child_chain_utxohash = ?)", "standard_exit", txoutput.child_chain_utxohash)),
+      order_by: [asc: :blknum, asc: :txindex, asc: :oindex]
+    )
+  end
+
+  @spec get_utxos(keyword) :: OMG.Utils.Paginator.t()
+  def get_utxos(params) do
+    address = Keyword.get(params, :address)
+    paginator = Paginator.from_constraints(params, @default_get_utxos_limit)
+    %{limit: limit, page: page} = paginator.data_paging
+    offset = (page - 1) * limit
+
+    from(query_get_utxos(address), limit: ^limit, offset: ^offset)
+    |> Repo.all()
```

Comment:
Hmm...can you re-describe the original issue again? I remember it is that when UTXO size become large the api would crash.

However, this pagination is still getting all utxos to memory and then do the pagination in memory. I would have strong doubt this solution would solve the issue or not though. (Also, I think the correct way of doing pagination with relational DB is to paginate on DB level....personally I would consider the Paginator here sth we should move away from)

[disclaimer: I haven't really touched relational db for a while 😛]

Reply:
@boolafish it is a DB sql `limit` and `offset` there. those are ecto queries. the paginator is just transforming the result.

Reply:
🤦‍♂️I see. sorry for the confusion.